File: Assignment 1/Solutions/Bio-Sequence Alignment.java
import java.util.HashMap;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        
        for (int testCase = 0; testCase < t; testCase++) {

            int length = scanner.nextInt();
            scanner.nextLine(); 
            String[] num = scanner.nextLine().split(" ");
            int sample = scanner.nextInt();
            scanner.nextLine(); 
            
            for (int s = 0; s < sample; s++) {

                String DNA = scanner.nextLine();
                boolean match = true;
                HashMap<String, Character> pair = new HashMap<>();
                
                if (DNA.length() != length) {

                    System.out.println("NO");
                    continue;
                }
                
                for (int i = 0; i < length; i++) {
                    
                    String currentNum = num[i];
                    char currentChar = DNA.charAt(i);
                    
                    if (!pair.containsKey(currentNum)) {
                        if (pair.containsValue(currentChar)) {
                            match = false;
                            break;
                        }
                        pair.put(currentNum, currentChar);
                    } else {
                        if (pair.get(currentNum) != currentChar) {
                            match = false;
                            break;
                        }
                    }
                }
                
                System.out.println(match ? "YES" : "NO");
            }
        }
        scanner.close();
    }
}

File: Assignment 1/Solutions/Bio-Sequence Alignment.py
t = int(input())

for _ in range(t):

    length = int(input())
    num = input().split()
    sample = int(input())

    for _ in range(sample):

        match = 1
        pair = {}
        DNA = input()
        
        if len(DNA) != length:
            print("NO")
            continue

        for i in range(length):

            if num[i] not in pair:
                if DNA[i] in pair.values():
                    match = 0
                    break
                pair[num[i]] = DNA[i]
             
            elif pair[num[i]] != DNA[i]:
                match = 0
                break

        if match == 1:
            print("YES")
        else:
            print("NO")
